### DISCUSSION TITLE
How to show only dabbrev but not code candidates when editing comments?

### DISCUSSION THREAD

**Discussion by IceAsteroid** [company-mode/company-mode#1498] [Q&A] [answered]
Hi, guys!

Company shows a list of candidates, including code completion candidates when editing comments in a programming mode, is it the default behavior, or because I misconfigured it?

In `emacs-lisp-mode` and `sh-mode`, specifically in `emacs-lisp-mode`, company shows dabbrev candiates only when editing comments or strings, in `sh-mode`, company doesn't  show completion popup when doing so.

I've set `eglot-stay-out-of '(company)`  in eglot.

For `sh-mode`, I'm using eglot + shellcheck.  

Here's the values of `completion-at-point-functions` & `company-backends` for these two modes:
```
;; In emacs-lisp-mode
;; completion-at-point-functions
(elisp-completion-at-point t)

;; company-backends
(company-bbdb company-semantic company-cmake company-capf company-clang company-files
              (company-dabbrev-code company-gtags company-etags company-keywords)
              company-oddmuse company-dabbrev)

;; In sh-mode
;; completion-at-point-functions
(eglot-completion-at-point sh-completion-at-point-function comint-completion-at-point) 
;; company-backends
((company-keywords company-capf company-files)
 (company-dabbrev-code company-gtags company-etags company-keywords)
 company-dabbrev)
```

But in `python-mode`, and `haskell-mode`, company shows code candidates in completion popup when editing comments.

I've tried pyright, pylsp, and jedi as lsp-server for python, and haskell-language-server for haskell, they all work like above, I don't know what component of the completion infrastructure determines this, so I asked here.  

Here's the values of `completion-at-point-functions` & `company-backends` in `python-ts-mode` & `haskell-mode`:
```
;; python-ts-mode
;; completion-at-point-functions
(eglot-completion-at-point python-completion-at-point t)
;; company-backends
(company-capf)

;; haskell-mode
;; completion-at-point-functions
(eglot-completion-at-point haskell-completions-sync-repl-completion-at-point haskell-completions-completion-at-point t)
;; company-backends
(company-bbdb company-semantic company-cmake company-capf company-clang company-files
              (company-dabbrev-code company-gtags company-etags company-keywords)
              company-oddmuse company-dabbrev)

```

Is there a way to generally show only dabbrev candidates when commenting in various programming modes?

**Comment by dgutov** [comment, marked as answer]
Hi! In company-mode (and this is similar when using other UIs) opting not to provide completions when inside comments is the responsibility of each individual.

Which is to say this should be a feature in Eglot. Here's the corresponding implementation in lsp-mode: https://github.com/emacs-lsp/lsp-mode/blob/78c194d2d3c34f14646a006574b9bb9369ea55cb/lsp-completion.el#L436-L437

To ask for a new feature in Eglot, it's recommended to use `M-x report-emacs-bug`.

**Reply by IceAsteroid** [reply]
Okay, I'll try contact to the Eglot team

**Comment by dgutov** [comment]
As a matter of personal configuration, though, you could also add an "around" advice for `company-capf--prefix` that depends on the return value of `(syntax-ppss)`.

**Reply by IceAsteroid** [reply]
Could you please describe how to utilize this function for the purpose a bit more?

I tried the following by utilizing to feed an empty string to `company-capf--prefix` to not have completion from other backends except from `dabbrev`, but it works not that well integrated

```
(defun my-company-capf-around (orig-fun &rest args)
  "Custom advice for `company-capf--prefix` to restrict completions in comments."
  (let ((syntax-info (syntax-ppss)))
    (if (nth 4 syntax-info)
        ""
      (apply orig-fun args))))

(advice-add 'company-capf--prefix :around #'my-company-capf-around)

```

I'll have to merge `company-dabbrev` into `company-capf` in order to make it work, which also shows dabbrev candidates when editing code. It'd be better if dabbrev candidates only show in editing comments 

This one as described above works
```
(setq company-backends '((company-capf
                          company-dabbrev
                          :separate)))
```


The following two ways are what I intended to use, but no success, the completion will not show when editing comments.

```
(setq company-backends '(company-capf
                         company-dabbrev
                         ))
```
or
```
(setq completion-at-point-functions '(eglot-completion-at-point
                                      python-completion-at-point
                                      dabbrev-capf))
```

**Reply by IceAsteroid** [reply]
I got some progress, but `dabbrev-capf` has a weird issue which is not caused by `company`. Just to share here.

I modified `eglot-completion-at-point` which is a backend in `completion-at-point-functions` instead of `company-capf--prefix` for finer tweaking. 

`dabbrev-capf` as a `completion-at-point-function` will output an error `Wrong type argument: stringp, nil`. If someone can help, please tell me.

Here's the code:

```
(defun my-eglot-capf-around-advice (orig-fun &rest args)
    (if (nth 4 (syntax-ppss))
        (funcall 'dabbrev-capf)
      (funcall orig-fun))
    )
(advice-add 'eglot-completion-at-point :around 'my-eglot-capf-around-advice)
```

**Reply by dgutov** [reply]
I don't recommend using `dabbrev-capf` here.

For the advice in your first example, return nil, not empty string:

```
(defun my-company-capf-around (orig-fun &rest args)
  "Custom advice for `company-capf--prefix` to restrict completions in comments."
  (let ((syntax-info (syntax-ppss)))
    (if (nth 4 syntax-info)
        nil
      (apply orig-fun args))))
      
(advice-add 'company-capf--prefix :around #'my-company-capf-around)
```

**Reply by IceAsteroid** [reply]
I've had a weirdo workaround that partially solved this dabbrev completion on comments. But I'll still post something to eglot team when I get sometime.

**Comment by IceAsteroid** [comment]
I'm so exited to tell that I finally got what I wanted!

The following configuration works great for my use cases, it may be as an example for someone, just to share it here.

Thank you for the amazing work, I'm grateful for it.

```
(defun my-fix/company-eglot-around-advice (orig-fun &rest args)
  ;; Only change backends for company & capf when eglot is in use in a buffer.
  (if (eglot-managed-p)
    (if (nth 4 (syntax-ppss))
        (let* ((company-backends '(company-dabbrev)))
          (funcall orig-fun)
          )
      (let* ((completion-at-point-functions '(eglot-completion-at-point))
             (company-backends '(company-capf
                                 company-dabbrev-code
                                 )))
        (funcall orig-fun)))
    (funcall orig-fun)
    ))
(advice-add 'company--perform :around 'my-fix/company-eglot-around-advice)

```


https://github.com/user-attachments/assets/771293cc-3bbd-49b8-89f2-07b3d0727f15
